---
title: "Initial analysis of data"
format: html
editor: visual
---

## Initial data analysis

Bring in packages

```{r}
pacman::p_load(rms, tidyverse, vroom, survival, lubridate, janitor, broom, Hmisc, gt, gtsummary, patchwork, data.table, survminer, qqman, ggforestplot)
```

Bring in covariates data: this includes covariates at the time of UKB attendance

Bring in outcome data:

```{r}
covars <- vroom("~/data/CRP/covariates_cleaned_rap.tsv.gz")
hes <- vroom("~/data/CRP/HES_icd_10_infections_only.tsv.gz")
prim_care <-  vroom("~/data/CRP/primary_care_icd_10_infections_only.tsv.gz")
metabs <- vroom("~/data/CRP/final_hdl_for_models.tsv.gz") %>%
  select(eid, 71:395)
prim_care_crp <- data.table::fread("~/data/CRP/primary_care_crp.tsv.gz")
common_prs_genome <- fread("~/data/CRP/gwas/genome_wide.score.gz")
#metabs_old <- vroom("~/data/CRP/final_hdl_for_models.tsv.gz") 
transplant <- vroom("~/data/CRP/transplant_list.tsv.gz")
common_prs <- fread("~/data/CRP/gwas/plink2.sscore.gz")

rare_var <- data.table::fread("~/data/CRP/exome/called_rare_var.raw.gz")

rare_var %>%
  filter(`1:159714024:G:A_A` == 2)
linker_IEU <- data.table::fread("~/data/CRP/ref_link/eids_merge.tsv.gz")
linker_MY <- data.table::fread("~/data/CRP/ref_link/linker_MY_APP.csv.gz")

prim_care_linked <- data.table::fread("~/data/CRP/icd_10_added_primary_care_events.tsv.gz") %>%
  select(eid) %>%
  distinct()


principle_components <- data.table::fread("~/data/CRP/principal_components.tsv.gz") %>%
  clean_names() %>%
  rename(eid = participant_id)

principle_components %>%
  select(-eid) %>%
  distinct()

colnames(principle_components) <- colnames(principle_components) %>%
  str_replace("genetic_principal_components_array_", "PC")


principle_components
quantile(covars$c_reactive_protein,0.0001, na.rm = T)
hes_crit <- fread("~/data/CRP/hesin_critical.txt.gz") %>%
  mutate(ccstartdate = dmy(ccstartdate),
         ccdisdate = dmy(ccdisdate))

```

Tidy genetic data first to keep neat

```{r}
common_prs <- common_prs %>%
  as_tibble() %>%
  select(ieu = IID, crp_prs = SCORE1_AVG) %>%
  mutate(crp_prs_scaled = scale(crp_prs)[,1]) %>%
  left_join(linker_MY) %>%
  select(eid, crp_prs, crp_prs_scaled)

colnames(rare_var)
rare_var <- rare_var %>%
  clean_names() %>%
  as_tibble() %>%
  select(participant_id = fid, contains("x1")) %>%
  mutate(participant_id = as.numeric(participant_id)) %>%
  select(rare_var_CRP = x1_159714024_g_a_a, participant_id) %>%
  left_join(linker_IEU) %>%
  drop_na(eid) %>%
  select(eid, rare_var_CRP)
common_prs_genome <- common_prs_genome %>%
  as_tibble() %>%
  select(ieu = IID, crp_prs_genome= SCORE1_AVG) %>%
  mutate(crp_prs_scaled_genome = scale(crp_prs_genome)[,1]) %>%
  left_join(linker_MY) %>%
  select(eid, crp_prs_genome, crp_prs_scaled_genome)


```

Generate a dataset for Cox regression: only including first event and filtering on after study.

```{r}

event_data <- covars %>%
  select(eid, study_entry) %>%
  left_join(hes) %>%
  mutate(event_dt = dmy(event_dt)) %>%
  arrange(event_dt) %>%
  filter(study_entry < event_dt) %>%
  mutate(time_event = as.numeric(difftime(event_dt,study_entry,  units = "days"))) %>%
  mutate(had_infection = if_else(time_event>0 ,1,0),
         had_sepsis = if_else(organ_system == "Sepsis",1,0),
         had_pneumonia = if_else(category == "Pneumonia",1,0),
         )
event_data %>%
  filter(str_detect(meaning, "J13"))

infection_events <- event_data %>%
  group_by(eid) %>%
  slice(1) %>%
  ungroup() %>%
  select(eid, time_event_infection = time_event, had_infection)

quantile(infection_events$time_event_infection)

sepsis_events <- event_data %>%
  filter(had_sepsis ==1) %>%
  group_by(eid) %>%
  slice(1) %>%
  ungroup() %>%
  select(eid, time_event_sepsis = time_event, had_sepsis)

quantile(sepsis_events$time_event_sepsis)

pneumonia_events <- event_data %>%
   filter(had_pneumonia == 1) %>%
  group_by(eid) %>%
  slice(1) %>%
  ungroup() %>%
  select(eid, time_event_pneumonia = time_event, had_pneumonia)

pneumococcal_events <- event_data %>%
     filter(str_detect(meaning, "J13")) %>%
  mutate(had_pneumococcus = 1) %>%
  group_by(eid) %>%
  slice(1) %>%
  ungroup() %>%
  select(eid, time_event_pneumococcus = time_event, had_pneumococcus)


pneumonia_minus_events <- event_data %>%
  filter(had_pneumonia == 1) %>%
     filter(!str_detect(meaning, "J13")) %>%
  mutate(had_pneumonia_minus = 1) %>%
  group_by(eid) %>%
  slice(1) %>%
  ungroup() %>%
  select(eid, time_event_pneumonia_minus = time_event, had_pneumonia_minus)

pneumococcal_events

quantile(pneumonia_events$time_event_pneumonia)


```

Link to covars;

1.  generate cox regression df
2.  generation time to death
3.  death censor

```{r}


cox_ready <- covars %>%
  left_join(infection_events) %>%
  left_join(sepsis_events) %>%
  left_join(pneumonia_events) %>%
  left_join(pneumococcal_events) %>%
  left_join(pneumonia_minus_events) %>%
  replace_na(list(had_pneumonia =0, had_sepsis = 0, had_infection = 0, had_pneumococcus = 0, had_pneumonia_minus = 0)) %>%
   mutate(max_fu = as.numeric(difftime(ymd("2020-01-01"),study_entry,  units = "days"))) %>%
  mutate(time_to_death = as.numeric(difftime(date_of_death, study_entry, units = "days"))) %>%
  mutate(max_fu = if_else(is.na(time_to_death), max_fu, time_to_death)) %>%
    mutate(time_event_infection = if_else(is.na(time_event_infection), max_fu, time_event_infection)) %>%
  mutate(time_event_pneumonia = if_else(is.na(time_event_pneumonia), max_fu, time_event_pneumonia)) %>%
  mutate(time_event_sepsis = if_else(is.na(time_event_sepsis), max_fu, time_event_sepsis)) %>%
    mutate(time_event_pneumococcus = if_else(is.na(time_event_pneumococcus), max_fu, time_event_pneumococcus)) %>%
    mutate(time_event_pneumonia_minus = if_else(is.na(time_event_pneumonia_minus), max_fu, time_event_pneumonia_minus)) 
#generate death data

cox_ready <- cox_ready %>%
  mutate(pneumonia_death = case_when(
    had_pneumonia == 1 &
      time_event_pneumonia + 28 > time_to_death ~ 1,
    TRUE ~ 0
  )) %>%
   mutate(infection_death = case_when(
    had_infection == 1 &
      time_event_infection + 28 > time_to_death ~ 1,
    TRUE ~ 0
  )) %>%
   mutate(sepsis_death = case_when(
    had_sepsis == 1 &
      time_event_sepsis + 28 > time_to_death ~ 1,
    TRUE ~ 0
   )) 
  # %>%
  # mutate(time_event_infection_censored= if_else(is.na(date_of_death), time_event_infection, time_to_death)) %>%
  # mutate(time_event_sepsis_censored= if_else(is.na(date_of_death), time_event_sepsis, time_to_death)) %>%
  # mutate(time_event_pneumonia_censored= if_else(is.na(date_of_death), time_event_pneumonia, time_to_death)) 
  # 

#Generate censored death date

cox_ready <- cox_ready %>%
    mutate(time_event_death_censor = if_else(is.na(time_to_death), max_fu, time_to_death))







```

Table S1

```{r}
cox_ready %>%
  drop_na(c_reactive_protein) %>%
  mutate(c_reactive_protein = Hmisc::cut2(c_reactive_protein, g = 4)) %>%
  select(age_at_entry, sex, history_stroke, history_MI, history_asthma, history_copd, history_dementia, history_liver, history_cancer, ever_smoked, townsend_deprivation_index_at_recruitment, body_mass_index_bmi, creatinine, c_reactive_protein) %>%
    tbl_summary(by = "c_reactive_protein") %>%
  add_p()
    

```

Figure 1

```{r}


cox_ready %>%
  mutate(crp = ntile(log(c_reactive_protein), 20)) %>%
  group_by(crp) %>%
     drop_na(c_reactive_protein) %>%
  group_by(crp) %>%
  summarise(median_crp = median(c_reactive_protein),
            bmean_pneumonia = mean(had_infection),
            n = n()) %>%
  gt::gt()

p1<-cox_ready %>%
  mutate(crp = ntile(log(c_reactive_protein), 10)) %>%
  group_by(crp) %>%
     drop_na(c_reactive_protein) %>%
  group_by(crp) %>%
  summarise(median_crp = median(c_reactive_protein),
            bmean_pneumonia = mean(had_infection)) %>%
   pivot_longer(contains("mean")) %>%
   ggplot(aes(x = median_crp, y = value)) +
   geom_point() +
  geom_line() +
   theme_bw() +
   ylab("Proportion with infection") +
    xlab("CRP (mg/L)") +
  xlim(0.25,9) +
  ylim(0.1,0.22) 


ggsave("Figure_1.tiff", p1, height = 3, width = 6, compression = "lzw")


p.s1<-cox_ready %>%
  mutate(crp = ntile(log(c_reactive_protein), 10)) %>%
  group_by(crp) %>%
     drop_na(c_reactive_protein) %>%
  group_by(crp) %>%
  summarise(median_crp = median(c_reactive_protein),
            bmean_pneumonia = mean(had_pneumonia),
            bmean_sepsis = mean(had_sepsis),
            ) %>%
   pivot_longer(contains("mean")) %>%
   ggplot(aes(x = median_crp, y = value, colour = name)) +
   geom_point() +
  geom_line() +
   theme_bw() +
   ylab("Proportion with infection") +
    xlab("CRP (mg/L)") +
  scale_color_discrete(name = "", labels = c("Pneumonia", "Sepsis")) +
  theme(legend.position = "bottom")

ggsave("Figure_S1.tiff", p.s1, height = 5, width = 5, compression = "lzw")
```

Add in and scale variables:

```{r}


final <- cox_ready %>%
  # bind_cols(scaled, log_scaled) %>%
  # left_join(metab_model) %>%
  mutate(ln_crp = log(c_reactive_protein)) 

m_resid <- lm(ln_crp ~ body_mass_index_bmi, data = final)
tidy(m_resid)




```

Data in tables

```{r}

cox_ready %>%
  filter(had_pneumonia == 1) %>%
  count(pneumonia_death) %>%
  mutate(perc = n/sum(n))

```

Run multivariable models:

table 2

```{r}
x <- "Surv(time_event_infection, had_infection)" 

run_table_2 <- function(x, y) {
f <- as.formula(sprintf(paste0(x, "~ ln_crp + sex + rcs(age_at_entry) + uk_biobank_assessment_centre + history_stroke + history_MI + history_copd + history_liver + history_cancer + ever_smoked + creatinine ")))
m <- (coxph(formula = f, data = final))
m1 <- tidy(m, exponentiate = T, conf.int = T) %>% mutate(model = y) %>% mutate(adj = "nobmi")

f <- as.formula(sprintf(paste0(x, "~ ln_crp + sex + rcs(age_at_entry) + uk_biobank_assessment_centre + history_stroke + history_MI + history_copd + history_liver + history_cancer + ever_smoked + creatinine + body_mass_index_bmi ")))
m <- (coxph(formula = f, data = final))
m2 <- tidy(m, exponentiate = T, conf.int = T) %>% mutate(model = y) %>% mutate(adj = "bmi")
m1 %>%
  bind_rows(m2)
}

tab2 <- bind_rows(
run_table_2("Surv(time_event_infection, had_infection)", "Infection"),
run_table_2("Surv(time_event_pneumonia, had_pneumonia)", "Pneumonia"),
run_table_2("Surv(time_event_sepsis, had_sepsis)", "Sepsis"),

run_table_2("Surv(time_event_death_censor, infection_death)", "Infection death"),
run_table_2("Surv(time_event_death_censor, pneumonia_death)", "Pneumonia death"),
run_table_2("Surv(time_event_death_censor, sepsis_death)", "Sepsis death")

)


tab2 %>%
  filter(term == "ln_crp") %>%
  mutate(across(c(estimate,conf.low, conf.high, p.value),~ signif(.x, digits = 3))) %>%
  transmute(model, estimate = paste0(estimate, " (", conf.low, "-",conf.high,")"), p.value, adj) %>%
  group_by(model) %>%
  gt::gt()
```

```{r}




run_table_2 <- function(x, y) {
f <- as.formula(sprintf(paste0(x, "~ ln_crp + sex + rcs(age_at_entry) + uk_biobank_assessment_centre + history_stroke + history_MI + history_copd + history_liver + history_cancer + ever_smoked + creatinine ")))
m <- (coxph(formula = f, data = trans_cohort))
m1 <- tidy(m, exponentiate = T, conf.int = T) %>% mutate(model = y) %>% mutate(adj = "nobmi")

m1
}

tab2 <- bind_rows(
run_table_2("Surv(time_event_infection, had_infection)", "Infection"),
run_table_2("Surv(time_event_pneumonia, had_pneumonia)", "Pneumonia"),
run_table_2("Surv(time_event_sepsis, had_sepsis)", "Sepsis"),
run_table_2("Surv(time_event_death_censor, pneumonia_death)", "Pneumonia death"),


)


tab2 %>%
  filter(term == "ln_crp") %>%
  mutate(across(c(estimate,conf.low, conf.high, p.value),~ signif(.x, digits = 3))) %>%
  transmute(model, estimate = paste0(estimate, " (", conf.low, "-",conf.high,")"), p.value,) %>%
  group_by(model) %>%
  gt::gt()

#infection

f <- as.formula(sprintf("Surv(time_event_infection, had_infection) ~ rcs(ln_crp, 3) + sex + rcs(age_at_entry) + uk_biobank_assessment_centre + history_stroke + history_MI + history_copd + history_liver + history_cancer + ever_smoked + creatinine + body_mass_index_bmi " ))
m <- (coxph(formula = f, data = final))
g1 = termplot(m, term = 1, se = T,plot = F)
g1 <- g1$ln_crp %>% mutate(model = "Infection")

f <- as.formula(sprintf("Surv(time_event_pneumonia, had_pneumonia) ~ rcs(ln_crp, 3) + sex + rcs(age_at_entry) + uk_biobank_assessment_centre + history_stroke + history_MI + history_copd + history_liver + history_cancer + ever_smoked + creatinine + body_mass_index_bmi" ))
m <- (coxph(formula = f, data = final))
g2 = termplot(m, term = 1, se = T,plot = F)
g2 <- g2$ln_crp %>% mutate(model = "Pneumonia")

f <- as.formula(sprintf("Surv(time_event_sepsis, had_sepsis) ~ rcs(ln_crp, 3) + sex + rcs(age_at_entry) + uk_biobank_assessment_centre + history_stroke + history_MI + history_copd + history_liver + history_cancer + ever_smoked + creatinine + body_mass_index_bmi" ))
m <- (coxph(formula = f, data = final))
g3 = termplot(m, term = 1, se = T,plot = F)
g3 <- g3$ln_crp %>% mutate(model = "Sepsis")




p.s2 <- g1 %>%
  bind_rows(g2, g3) %>%
  as_tibble() %>%
  mutate(lci = exp(y-se*1.96), uci = exp(y+se*1.96), y = exp(y), x = exp(x)) %>%
  ggplot(aes(x = x, y = y)) +
  geom_line(aes(colour = model)) +
  geom_ribbon(aes(ymin = lci, ymax = uci, fill = model), alpha = 0.3) +
  scale_x_continuous(trans="log10") +
  scale_y_continuous(trans="log2") +
  xlab("C reactive protein (mg/l)") +
  ylab("Hazard Ratio (HR)") +
  theme_bw() +
  theme(legend.position = "bottom")



ggsave("Figure_S2.tiff", p.s2, height = 6, width = 6.5, compression = "lzw") 
```

Transplant

```{r}

trans_cohort_tbl <- cox_ready %>%
  select(eid, study_entry) %>%
  left_join(transplant) %>%
  filter(study_entry > transplant_date) %>%
  right_join(cox_ready)

cox_ready$c_reactive_protein

trans_cohort_tbl %>%
  drop_na(c_reactive_protein) %>%
  mutate(has_transplant = if_else(is.na(transplant_date), "no SOT", "SOT")) %>%
  select(age_at_entry, sex, history_stroke, history_MI, history_asthma, history_copd, history_dementia, history_liver, history_cancer, ever_smoked, townsend_deprivation_index_at_recruitment, body_mass_index_bmi, creatinine, c_reactive_protein, has_transplant) %>%
  tbl_summary(by = "has_transplant") %>%
  add_p()

trans_cohort<- final %>%
  select(eid, study_entry) %>%
  left_join(transplant) %>%
  filter(study_entry > transplant_date) %>%
  left_join(final) %>%
  mutate(time_to_transplant = difftime(transplant_date, study_entry))





run_multivariate_models <- function(x) {

  
  # f <- as.formula(sprintf("Surv(time_event_pneumonia, had_pneumonia) ~ %s", x))
 f <- as.formula(sprintf("Surv(time_event_infection, had_infection) ~ sex + age_at_entry + uk_biobank_assessment_centre + history_stroke + history_MI + history_copd + history_liver + history_cancer + ever_smoked + creatinine + %s" , x))
# f <- as.formula(sprintf("Surv(time_event_infection, had_infection) ~ %s", x))
m <- (coxph(formula = f, data = trans_cohort))
d <- summary(m)
tidy(m, conf.int = T, exponentiate = T) %>%
  mutate(rsq = d$rsq[1])
}

multi_v_res <- map_dfr("ln_crp", run_multivariate_models)
multi_v_res %>%
  filter(term == "ln_crp")
multi_v_res <- map_dfr("ln_crp + body_mass_index_bmi", run_multivariate_models)

trans_cohort$had_pneumonia


f <- as.formula(sprintf("Surv(time_event_infection, had_infection) ~ rcs(ln_crp, 3) + sex + rcs(age_at_entry) + uk_biobank_assessment_centre + history_stroke + history_MI + history_copd + history_liver + history_cancer + ever_smoked + creatinine + body_mass_index_bmi " ))
m <- (coxph(formula = f, data = trans_cohort))
g1 = termplot(m, term = 1, se = T,plot = F)
g1 <- g1$ln_crp %>% mutate(model = "Infection")

f <- as.formula(sprintf("Surv(time_event_pneumonia, had_pneumonia) ~ rcs(ln_crp, 3) + sex + rcs(age_at_entry) + uk_biobank_assessment_centre + history_stroke + history_MI + history_copd + history_liver + history_cancer + ever_smoked + creatinine + body_mass_index_bmi" ))
m <- (coxph(formula = f, data = trans_cohort))
g2 = termplot(m, term = 1, se = T,plot = F)
g2 <- g2$ln_crp %>% mutate(model = "Pneumonia")

f <- as.formula(sprintf("Surv(time_event_sepsis, had_sepsis) ~ rcs(ln_crp, 3) + sex + rcs(age_at_entry) + uk_biobank_assessment_centre + history_stroke + history_MI + history_copd + history_liver + history_cancer + ever_smoked + creatinine + body_mass_index_bmi" ))
m <- (coxph(formula = f, data = trans_cohort))
g3 = termplot(m, term = 1, se = T,plot = F)
g3 <- g3$ln_crp %>% mutate(model = "Sepsis")




p.s3 <- g1 %>%
  bind_rows(g2, g3) %>%
  as_tibble() %>%
  mutate(lci = exp(y-se*1.96), uci = exp(y+se*1.96), y = exp(y), x = exp(x)) %>%
  ggplot(aes(x = x, y = y)) +
  geom_line(aes(colour = model)) +
  geom_ribbon(aes(ymin = lci, ymax = uci, fill = model), alpha = 0.3) +
  scale_x_continuous(trans="log10") +
  scale_y_continuous(trans="log2") +
  xlab("C reactive protein (mg/l)") +
  ylab("Hazard Ratio (HR)") +
  theme_bw() +
  theme(legend.position = "bottom")


p.s2


ggsave("Figure_S2.tiff", p.s2, height = 6, width = 6.5, compression = "lzw") 
```

1.  primary care

```{r}
crp_only <- cox_ready %>%
   select(eid, c_reactive_protein, study_entry)

prim_care_crp %>%
   filter(crp_gp != 0) %>%
   filter(crp_gp != 5) %>%
   left_join(crp_only) %>%
   mutate(distance = abs(as.numeric(difftime(event_dt_crp, study_entry, units = "days")))) %>%
   arrange(distance) %>%
   group_by(eid) %>%
   slice(1) %>%
   ungroup() %>%
   mutate(distance = cut2(distance, cuts = c(2,10,100))) %>%
   drop_na(distance) %>%
   ggplot(aes(x = (crp_gp), y = (c_reactive_protein))) +
    geom_point() +
    geom_smooth(method = "lm") +
   facet_wrap(~ distance) +
   theme_bw() +
   scale_x_log10() +
   scale_y_log10() +
   xlab("Primary care CRP") +
  ylab("CRP measured at biobank")



prim_care_crp %>%
  filter(crp_gp != 0) %>%
   filter(crp_gp != 5) %>%
   left_join(crp_only) %>%
   mutate(distance = abs(as.numeric(difftime(event_dt_crp, study_entry, units = "days")))) %>%
  arrange(distance) %>%   
  group_by(eid) %>%
   slice(1) %>%
   ungroup() %>%
   mutate(distance = cut2(distance, cuts = c(2,10,100))) %>%
   drop_na(distance) %>%
   group_by(distance) %>%
   summarise(cor_c = cor(crp_gp, c_reactive_protein, use = "complete.obs"))



```

Generate data

```{r}
set.seed(123)
random_choice <- prim_care_crp %>%
   filter(crp_gp != 0) %>%
  filter(crp_gp != 5) %>%
   left_join(crp_only) %>%
  filter(event_dt_crp > study_entry) %>%
   group_by(eid) %>%
   sample_n(1)



event_data <- random_choice %>%
  select(eid, study_entry =  event_dt_crp) %>%
  left_join(hes) %>%
  mutate(event_dt = dmy(event_dt)) %>%
  arrange(event_dt) %>%
  filter(study_entry < event_dt) %>%
  mutate(time_event = as.numeric(difftime(event_dt,study_entry,  units = "days"))) %>%
  mutate(had_infection = if_else(time_event>0 ,1,0),
         had_sepsis = if_else(organ_system == "Sepsis",1,0),
         had_pneumonia = if_else(category == "Pneumonia",1,0),
         )

infection_events <- event_data %>%
  group_by(eid) %>%
  slice(1) %>%
  ungroup() %>%
  select(eid, time_event_infection = time_event, had_infection)

quantile(infection_events$time_event_infection)

sepsis_events <- event_data %>%
  filter(had_sepsis ==1) %>%
  group_by(eid) %>%
  slice(1) %>%
  ungroup() %>%
  select(eid, time_event_sepsis = time_event, had_sepsis)

quantile(sepsis_events$time_event_sepsis)

pneumonia_events <- event_data %>%
  filter(had_pneumonia ==1) %>%
  group_by(eid) %>%
  slice(1) %>%
  ungroup() %>%
  select(eid, time_event_pneumonia = time_event, had_pneumonia)

quantile(pneumonia_events$time_event_pneumonia)
```

generate primary care cox

```{r}


cox_ready_primary_care<- covars %>%
  inner_join(random_choice) %>%
  left_join(infection_events) %>%
  left_join(sepsis_events) %>%
  left_join(pneumonia_events) %>%
  replace_na(list(had_pneumonia =0, had_sepsis = 0, had_infection = 0)) %>%
   mutate(max_fu = as.numeric(difftime(ymd("2020-12-31"),study_entry,  units = "days"))) %>%
  mutate(time_event_infection = if_else(is.na(time_event_infection), max_fu, time_event_infection)) %>%
  mutate(time_event_pneumonia = if_else(is.na(time_event_pneumonia), max_fu, time_event_pneumonia)) %>%
  mutate(time_event_sepsis = if_else(is.na(time_event_sepsis), max_fu, time_event_sepsis)) %>%
  mutate(time_to_death = as.numeric(difftime(date_of_death, study_entry, units = "days"))) %>%
  mutate(time_event_infection= if_else(is.na(date_of_death), time_event_infection, time_to_death)) %>%
  mutate(time_event_sepsis= if_else(is.na(date_of_death), time_event_sepsis, time_to_death)) %>%
  mutate(time_event_pneumonia= if_else(is.na(date_of_death), time_event_pneumonia, time_to_death)) %>%
  mutate(ln_crp = log(crp_gp))


cox_ready_primary_care
```

```{r}

run_table_2 <- function(x, y) {
f <- as.formula(sprintf(paste0(x, "~ ln_crp + sex + rcs(age_at_entry) + uk_biobank_assessment_centre + history_stroke + history_MI + history_copd + history_liver + history_cancer + ever_smoked + creatinine ")))
m <- (coxph(formula = f, data = cox_ready_primary_care))
m1 <- tidy(m, exponentiate = T, conf.int = T) %>% mutate(model = y) %>% mutate(adj = "nobmi")

f <- as.formula(sprintf(paste0(x, "~ ln_crp + sex + rcs(age_at_entry) + uk_biobank_assessment_centre + history_stroke + history_MI + history_copd + history_liver + history_cancer + ever_smoked + creatinine + body_mass_index_bmi ")))
m <- (coxph(formula = f, data = cox_ready_primary_care))
m2 <- tidy(m, exponentiate = T, conf.int = T) %>% mutate(model = y) %>% mutate(adj = "bmi")
m1 %>%
  bind_rows(m2)
}

tab2 <- bind_rows(
run_table_2("Surv(time_event_infection, had_infection)", "Infection"),
run_table_2("Surv(time_event_pneumonia, had_pneumonia)", "Pneumonia"),
run_table_2("Surv(time_event_sepsis, had_sepsis)", "Sepsis"))

tab2 %>%
  filter(term == "ln_crp") %>%
  mutate(across(c(estimate,conf.low, conf.high, p.value),~ signif(.x, digits = 3))) %>%
  transmute(model, estimate = paste0(estimate, " (", conf.low, "-",conf.high,")"), p.value, adj) %>%
  group_by(model) %>%
  gt::gt()
```

```{r}

f <- as.formula(sprintf("Surv(time_event_infection, had_infection) ~ rcs(ln_crp) + sex + rcs(age_at_entry) + uk_biobank_assessment_centre + history_stroke + history_MI + history_copd + history_liver + history_cancer + ever_smoked + creatinine + rcs(body_mass_index_bmi)" ))
m <- (coxph(formula = f, data = final))
g1 = termplot(m, term = 1, se = T,plot = F)
g1 <- g1$ln_crp %>% mutate(model = "Infection")

f <- as.formula(sprintf("Surv(time_event_pneumonia, had_pneumonia) ~ rcs(ln_crp) + sex + rcs(age_at_entry) + uk_biobank_assessment_centre + history_stroke + history_MI + history_copd + history_liver + history_cancer + ever_smoked + creatinine + rcs(body_mass_index_bmi)" ))
m <- (coxph(formula = f, data = final))
g2 = termplot(m, term = 1, se = T,plot = F)
g2 <- g2$ln_crp %>% mutate(model = "Pneumonia")

f <- as.formula(sprintf("Surv(time_event_sepsis, had_sepsis) ~ rcs(ln_crp) + sex + rcs(age_at_entry) + uk_biobank_assessment_centre + history_stroke + history_MI + history_copd + history_liver + history_cancer + ever_smoked + creatinine + rcs(body_mass_index_bmi)" ))
m <- (coxph(formula = f, data = final))
g3 = termplot(m, term = 1, se = T,plot = F)
g3 <- g3$ln_crp %>% mutate(model = "Sepsis")




g1 %>%
  bind_rows(g2, g3) %>%
  as_tibble() %>%
  filter(x >-0.5) %>%
  mutate(lci = exp(y-se*1.96), uci = exp(y+se*1.96), y = exp(y), x = exp(x)) %>%
  ggplot(aes(x = x, y = y)) +
  geom_line(aes(colour = model)) +
  geom_ribbon(aes(ymin = lci, ymax = uci, fill = model), alpha = 0.3) +
  scale_y_continuous(trans="log2") +
  xlab("C reactive protein (mg/l)") +
  ylab("Hazard Ratio (HR)") +
  theme_bw() +
  scale_x_continuous(trans="log10") 
```

Rare variant

```{r}




rare_final <- rare_var %>%
  left_join(final) %>%
  drop_na(rare_var_CRP) %>%
  left_join(principle_components)

m <-lm(ln_crp ~ rare_var_CRP, data = rare_final)
tidy(m)

rare_final %>%
  count(rare_var_CRP) %>%
  mutate(x = n/sum(n) * 100)

rare_final %>%
  drop_na(c_reactive_protein) %>%
  group_by(rare_var_CRP) %>%
  summarise(median = median(c_reactive_protein, na.rm = T),
            upper = quantile(c_reactive_protein, 0.75),
            lower = quantile(c_reactive_protein, 0.25), n = n())

p2 <- rare_final %>%
  mutate(rs77832441_A = if_else(rare_var_CRP == 1, "Carrier", "Non Carrier")) %>%
  select(eid, rs77832441_A, c_reactive_protein) %>%
  ggplot(aes(x = c_reactive_protein)) +
  geom_density(aes(group = rs77832441_A, color = rs77832441_A, fill =rs77832441_A), alpha = 0.3) +
  scale_x_log10() +
  xlab("CRP (mg/L)") +
  theme_bw() +
  theme(legend.position = "bottom")

p2
ggsave("Figure_2.tiff", p2, compression = "lzw", height = 5, width = 6)
rare_final %>%
  count(rare_var_CRP)


rare_final %>%
  inner_join(prim_care_crp) %>%
  # group_by(eid) %>%
  # slice(1) %>%
  # ungroup() %>%
  group_by(rare_var_CRP) %>%
  summarise(median = mean(crp_gp, na.rm = T), n = n())


rare_final %>%
  inner_join(prim_care_crp) %>%
  group_by(eid) %>%
  slice(1) %>%
  ungroup() %>%
  group_by(rare_var_CRP) %>%
  summarise(median = mean(crp_gp, na.rm = T), n = n())

rare_final %>%
  filter(rare_var_CRP == 2) %>%
  inner_join(prim_care_crp) %>%
  count(crp_gp)

rare_final %>%
  filter(rare_var_CRP == 2) %>%
  inner_join(prim_care_crp) %>%
   group_by(eid) %>%
   slice(1) %>%
   ungroup() %>%
  group_by(rare_var_CRP) %>%
  summarise(median = median(crp_gp, na.rm = T), n = n())





```

```{r}


# 
rare_final_no_hom <- rare_final %>%
   filter(rare_var_CRP != 2 )

x <- "Surv(time_event_infection, had_infection)" 

run_table_gen <- function(x, y) {


f <- as.formula(sprintf(paste0(x, "~ as.factor(rare_var_CRP) + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + age_at_entry + sex ")))
m <- (coxph(formula = f, data = rare_final_no_hom))
m2 <- tidy(m, exponentiate = T, conf.int = T) %>% mutate(model = y) 
return(m2)
}

tab2 <- bind_rows(
run_table_gen("Surv(time_event_infection, had_infection)", "Infection"),
run_table_gen("Surv(time_event_pneumonia, had_pneumonia)", "Pneumonia"),
run_table_gen("Surv(time_event_pneumonia_minus, had_pneumonia_minus)", "Pneumonia without pneumococcus"),
run_table_gen("Surv(time_event_sepsis, had_sepsis)", "Sepsis"),
run_table_gen("Surv(time_event_pneumococcus, had_pneumococcus)", "Pneumonia due to Streptococcus pneumoniae"),


run_table_gen("Surv(time_event_death_censor, infection_death)", "Infection death"),
run_table_gen("Surv(time_event_death_censor, pneumonia_death)", "Pneumonia death"),
run_table_gen("Surv(time_event_death_censor, sepsis_death)", "Sepsis death")

)




tab2 %>%
  filter(str_detect(term, "rare_var_CRP")) %>%
  mutate(across(c(estimate,conf.low, conf.high, p.value),~ signif(.x, digits = 3))) %>%
  transmute(model, estimate = paste0(estimate, " (", conf.low, "-",conf.high,")"), p.value, term) %>%
  group_by(model) %>%
  gt::gt()







fit <- survfit(Surv(time_event_infection, had_infection) ~ rare_var_CRP, data = rare_final_no_hom)

p3.3 <- ggsurvplot(fit, data = rare_final_no_hom, conf.int = T, ylim = c(0.85,1),
                   pval = T,
                    xscale = "d_y", # convert time as days to years in x-axis
                    break.x.by = 365.25, # break X axis in time intervals by years as integers
                xlab = "Time (years)",
                xlim = c(0,3560),
                 pval.coord = c(0, .87), # p-value location
                pval.size = 6.5, # adjust p-value size
               legend.title = "", legend.labs = c("wild type", "heterozygote"), legend = "bottom",
               ylab ="Survival without infection",
               title = "Incidence of infection")


p3.3
fit <- survfit(Surv(time_event_pneumonia, had_pneumonia) ~ rare_var_CRP, data = rare_final_no_hom)

p3.4 <- ggsurvplot(fit, data = rare_final_no_hom, conf.int = T, ylim = c(0.95,1),
                   pval = T,
                    xscale = "d_y", # convert time as days to years in x-axis
                    break.x.by = 365.25, # break X axis in time intervals by years as integers
                xlab = "Time (years)",
                xlim = c(0,3560),
                 pval.coord = c(0, .960), # p-value location
                pval.size = 6.5, # adjust p-value size
               legend.title = "", legend.labs = c("wild type", "heterozygote"), legend = "bottom",
               ylab ="Survival without pneumonia",
               title = "Incidence of pneumonia")

p3.4
tab2 %>%
  filter(str_detect(term, "rare_var_CRP")) %>%
  mutate(estimate = log(estimate)) %>%
  arrange(estimate) %>%
  filter(!str_detect(model, "death")) %>%
  ggforestplot::forestplot(name = model, estimate = estimate, se = std.error, logodds = T)
  


manifest <- fread("~/data/CRP/finngen/R8_manifest.tsv")
manifest <- manifest %>%
  select(condition = phenocode, name, num_cases, num_controls)
finngen <- fread("~/data/CRP/finngen/crp_finngen_rare_var_only.tab.gz")
finngen2 <-finngen %>%
  as_tibble() %>%
  select(pval = V7,
         beta = V9, 
         sebeta = V10,
         af_alt_cases = V12, 
         condition = V14) %>%
  mutate(condition = str_remove(condition, "finngen_R8_")) %>%
  left_join(manifest) %>%
  filter(af_alt_cases >0) %>% 
  filter(num_cases >100) %>%
  mutate(fdrp = p.adjust(pval, method = "fdr")) %>%
  mutate(cases = af_alt_cases * num_cases)  %>%
  arrange(pval) 



infections <- manifest %>%
  filter(str_detect(condition, "AB1") | str_detect(name, "Cystitis|Cellulitis|Appendicitis|Meningitis|Endocard")) 


finngen2 %>%
  right_join(infections) %>%
  arrange(pval) %>%
  transmute(name,
            num_cases,
            num_controls,
    odds_ratio = exp(beta),
     lower_95 = exp(beta - 1.96*sebeta),
         upper_95 = exp(beta + 1.96*sebeta),
        
    pval,
    af_alt_cases
    ) %>%
  gt::gt()
ukb_rare <- tab2 %>%
  filter(str_detect(term, "rare_var_CRP")) %>%
  mutate(estimate = log(estimate)) %>%
  arrange(estimate) %>%
  filter(!str_detect(model, "death")) %>%
  select(name = model,
         sebeta = std.error,
         beta = estimate) %>%
  mutate(cohort = "UK Biobank")




p3.1 <- tab2 %>%
  filter(str_detect(term, "rare_var_CRP")) %>%
  mutate(estimate = log(estimate)) %>%
  arrange(estimate) %>%
  filter(!str_detect(model, "death")) %>%
  select(name = model,
         sebeta = std.error,
         beta = estimate) %>%
  mutate(cohort = "UK Biobank")  %>%
  forestplot(name = name,
             estimate = beta,
             se = sebeta,
             logodds = T) +
  xlab("Hazard ratio (95% CI)") +
  ggtitle("UK Biobank")

finngen2 %>%
  arrange(beta) %>%
  filter(str_detect(condition, "J10_LOWE|PULM_INFECTIONS|AB1_BACTINF_NOS|J10_PNEUMOP|J10_PNEUMOB|SEPSI")) %>%
  mutate(cohort = "FinnGen") %>%
  bind_rows(ukb_rare) %>%
  mutate(group = as.factor(cohort)) %>%
  mutate(group = relevel(group, ref = "UK Biobank")) %>%
  forestplot(name = name,
             estimate = beta,
             se = sebeta,
             logodds = T) +
  ggforce::facet_col(
    facets = ~group,
    scales = "free_y",
    space = "free"
  )


p3.2 <- finngen2 %>%
  arrange(beta) %>%
  filter(str_detect(condition, "J10_LOWE|PULM_INFECTIONS|AB1_BACTINF_NOS|J10_PNEUMOP|J10_PNEUMOB|SEPSI")) %>%
  mutate(cohort = "FinnGen") %>%

  forestplot(name = name,
             estimate = beta,
             se = sebeta,
             logodds = T) +
  xlab("Odds ratio(95% CI)") +
  ggtitle("FinnGen")
event_data %>%
  filter(str_detect(meaning, "A403")) %>%
  count(meaning)
  

(p3.3$plot + p3.4$plot ) + plot_annotation(tag_levels = "A")
ggsave("Figure_3.tiff", compression = "lzw", height = 7, width = 14)

p3.1 + p3.2
ggsave("Figure_4.tiff", compression = "lzw", height = 5, width = 12)


hes <- vroom("~/data/CRP/hes_icd_added.txt.gz")
j13 <- hes %>%
  filter(meaning == "J13")

rare_final_j13 <- rare_final %>%
  mutate(j13 = if_else(eid %in% j13$eid,1,0)) 
 
m <- glm(j13 ~ rare_var_CRP + age_at_entry + sex + PC1 + PC2, data = rare_final_j13, family = "binomial")
tidy(m)
```

# MR

```{r}
conversions <- read_tsv("~/data/CRP/finngen/conversions.csv")
conversions
IL6R_start <- 154377819 - 3e5
IL6R_end <- 154441926 +3e5
tnf_start <- 6437923-3e5 
tnf_end <- 6451280 + 3e5
il1ra_start <- 113875548 -3e5
il1ra_end <- 113891591 + 3e5
crp_start <- 159682079 -3e5
crp_end <- 159684379 + 3e5
crp  <- data.table::fread("~/data/CRP/gwas/GCST90029070_buildGRCh37.tsv.gz") 

tnf_exposures <- crp %>%
  filter(chromosome == 12 & (base_pair_location > tnf_start) & (base_pair_location < tnf_end)) %>%
  select(SNP = variant_id, effect_allele.exposure = effect_allele, other_allele.exposure = other_allele, beta.exposure = beta, se.exposure = standard_error, pval.exposure = p_value) %>%
  as_tibble() %>%
  mutate(eaf.exposure = NA) %>%
  mutate(exposure ="TNF", id.exposure = "TNF") 

tnf_exposures_clumped <- tnf_exposures%>%
  mutate(pval.exposure = as.numeric(pval.exposure)) %>%
   filter(pval.exposure < 5e-8)  %>%
  clump_data(clump_r2 = 0.5)


il1ra <- crp %>%
  filter(chromosome == 2 & (base_pair_location > il1ra_start) & (base_pair_location < il1ra_end)) %>%
  select(SNP = variant_id, effect_allele.exposure = effect_allele, other_allele.exposure = other_allele, beta.exposure = beta, se.exposure = standard_error, pval.exposure = p_value) %>%
  as_tibble() %>%
  mutate(eaf.exposure = NA) %>%
  mutate(exposure ="IL1RA", id.exposure = "Il1RA") 

il1ra_clumped <- il1ra %>%
  mutate(pval.exposure = as.numeric(pval.exposure)) %>%
  filter(pval.exposure < 5e-8)  %>%
  clump_data(clump_r2 = 0.1)


il6r_clumped <- crp %>%
  filter(chromosome == 1 & (base_pair_location > IL6R_start) & (base_pair_location < IL6R_end)) %>%
  select(SNP = variant_id, effect_allele.exposure = effect_allele, other_allele.exposure = other_allele, beta.exposure = beta, se.exposure = standard_error, pval.exposure = p_value) %>%
  as_tibble() %>%
  mutate(eaf.exposure = NA) %>%
  mutate(exposure ="IL-6", id.exposure = "IL-6") %>%
  mutate(pval.exposure = as.numeric(pval.exposure)) %>%
  filter(pval.exposure < 5e-8)  %>%
  clump_data(clump_r2 = 0.1)


crp_clumped <- crp %>%
  filter(chromosome == 1 & (base_pair_location > crp_start) & (base_pair_location < crp_end)) %>%
  select(SNP = variant_id, effect_allele.exposure = effect_allele, other_allele.exposure = other_allele, beta.exposure = beta, se.exposure = standard_error, pval.exposure = p_value) %>%
  as_tibble() %>%
  mutate(eaf.exposure = NA) %>%
  mutate(exposure ="CRP", id.exposure = "CRP") %>%
  mutate(pval.exposure = as.numeric(pval.exposure)) %>%
  filter(pval.exposure < 5e-8)  %>%
  clump_data(clump_r2 = 0.01)

all_exposures <- bind_rows(tnf_exposures_clumped, il6r_clumped, crp_clumped, il1ra_clumped)

outcomes_r8 <- vroom("~/data/CRP/finngen/crp_finngen_whole_region_every.tab.gz", col_names= FALSE) %>%
  as_tibble() %>%
  select(SNP = X5,
         pval.outcome = X7,
         beta.outcome = X9,
         se.outcome = X10,
         eaf.outcome = X11,
         outcome = X14,
         effect_allele.outcome = X4,
         other_allele.outcome = X3) %>%
  mutate(id.outcome = outcome)



finngen_conversions <- conversions %>%
  mutate(id.outcome = str_remove(str_replace(id.outcome,"R6", "R8"), ".gz")) %>% filter(str_detect(id.outcome,"finngen"))

filtered_finngen <-outcomes_r8 %>%
  filter(id.outcome %in% finngen_conversions$id.outcome | str_detect(outcome, "J10_PNEUMOP")) 


ao <- available_outcomes() %>%
  filter(author == "Hamilton F")
outcomes_ukb <- extract_outcome_data(all_exposures$SNP, ao$id)
outcomes_r8 %>%
  count(outcome) %>%
  filter(str_detect(outcome, "J10_PNE"))
```

```{r}

outcomes <- bind_rows(outcomes_ukb, filtered_finngen)
dat <- harmonise_data(crp_clumped, filtered_finngen)
dat_ukb <- harmonise_data(crp_clumped, outcomes_ukb)



dat_pneumococcal <- harmonise_data(crp_clumped, pneumococcal)
res_pneumo <- mr(dat_pneumococcal, method = "mr_ivw") %>%
  mutate(name = "Pneumonia due to Streptococcus pneumoniae")
res <- mr(dat, method = "mr_ivw") %>%
  left_join(finngen_conversions)

res <- res %>%
  mutate(name = if_else(is.na(name), "Pneumonia due to Streptococcus pneumoniae", name))


res_ukb <- mr(dat_ukb, method = "mr_ivw") %>%
  filter(!str_detect(outcome, "death|crit")) %>%
  mutate(name = word(outcome, start = 1, end = 1)) %>%
  mutate(name = if_else(name == "Cholecystitits", "Cholecystitis", name))

res_ukb


res_total <- bind_rows(res, res_ukb) %>%
  filter(!str_detect(name, "Endo"))

res_total

ivw <- function(df) {
  TE = df$b
  SE = df$se
  m1 <- metagen(TE, SE) 
  tibble(estimate = m1$TE.fixed,
         std.error= m1$seTE.fixed,
         het = m1$pval.Q,
         pval = m1$pval.fixed,
         lower = m1$lower.fixed,
         upper = m1$upper.fixed)
  
}



p1 <- res_total %>%
  filter(exposure == "CRP") %>%
  filter(!name == "Streptococcal sepsis") %>%
    filter(!name == "Osteomyelitis") %>%
  filter(id.outcome !="ieu-b-5088") %>%
  group_by(name) %>%
  nest() %>%
  mutate(meta_an = map(data, ivw)) %>%
  unnest(meta_an) %>%
  ungroup() %>%
  select(name, b = estimate, se = std.error, pval)  %>%
  arrange(-b) %>%
  ggforestplot::forestplot(name = name, estimate = b, se = se, logodds = T) +
  theme_bw() +
  xlab("Odds ratio per log natural increase in CRP (95% CI)")


p1




res_total %>%
  filter(exposure == "cisCRP") %>%
    mutate(b = -b) %>%
  group_by(name) %>%
  nest() %>%
  mutate(meta_an = map(data, ivw)) %>%
  unnest(meta_an) %>%
  drop_na(het) %>%

  ungroup() %>%
  mutate(across(c(estimate,lower, upper), exp)) %>%
  view()
 
res_ukb %>%
  arrange(pval)
```

#Generate s pneumo GWAS

```{r}
hes <- vroom("~/data/CRP/hes_icd_added.txt.gz")
j13 <- hes %>%
  filter(meaning == "J13")
linker_MY %>%

  mutate(J13= if_else(eid %in% j13$eid,1,0)) %>%
    select(FID = ieu, IID = ieu, J13) %>%
  distinct(IID, .keep_all = T) %>%
  write_tsv("~/data/CRP/J13_pneumo.tsv.gz")
```

```{r}
hes <- vroom("~/data/CRP/hes_icd_added.txt.gz")




rare_final %>%
  filter(eid %in% prim_care_linked$eid) %>%
  select(eid, rare_var_CRP) %>%
  mutate(has_had_primary_care_infection = if_else(eid %in% prim_care$eid,1,0)) %>%
  count(rare_var_CRP, has_had_primary_care_infection) %>%
  group_by(rare_var_CRP) %>% mutate (perc = n /sum(n))
```

# Primary care sensitivity analyses

Generate a dataset for Cox regression: only including first event and filtering on after study.

```{r}

event_data <- covars %>%
  select(eid, study_entry) %>%
  left_join(prim_care) %>%
  mutate(event_dt = dmy(event_dt)) %>%
  arrange(event_dt) %>%
  filter(study_entry < event_dt) %>%
  mutate(time_event = as.numeric(difftime(event_dt,study_entry,  units = "days"))) %>%
  mutate(had_infection = if_else(time_event>0 ,1,0),
         had_sepsis = if_else(organ_system == "Sepsis",1,0),
         had_pneumonia = if_else(category == "Pneumonia",1,0),
         )

infection_events <- event_data %>%
  group_by(eid) %>%
  slice(1) %>%
  ungroup() %>%
  select(eid, time_event_infection = time_event, had_infection)

quantile(infection_events$time_event_infection)

sepsis_events <- event_data %>%
  filter(had_sepsis ==1) %>%
  group_by(eid) %>%
  slice(1) %>%
  ungroup() %>%
  select(eid, time_event_sepsis = time_event, had_sepsis)

quantile(sepsis_events$time_event_sepsis)

pneumonia_events <- event_data %>%
  filter(had_pneumonia ==1) %>%
  group_by(eid) %>%
  slice(1) %>%
  ungroup() %>%
  select(eid, time_event_pneumonia = time_event, had_pneumonia)

quantile(pneumonia_events$time_event_pneumonia)


all_events <- event_data %>%
  select(eid, time_event, had_infection) %>%
  arrange(eid) %>%
  group_by(eid) %>%
    mutate(tstart =lag(time_event)) %>% 
  ungroup() %>%
  mutate(tstart = if_else(is.na(tstart),0,tstart)) %>%
  distinct() 


max <- all_events %>%
  group_by(eid) %>%
  summarise(tstart = max(tstart))

d <- covars %>%
  filter(eid %in% prim_care_linked$eid) %>%
   mutate(max_fu = as.numeric(difftime(ymd("2020-12-31"),study_entry,  units = "days"))) %>%
  mutate(time_to_death = as.numeric(difftime(date_of_death, study_entry, units = "days"))) %>%
  mutate(max_fu = if_else(is.na(time_to_death), max_fu, time_to_death)) %>%
  select(time_event = max_fu, eid) %>%
  mutate(had_infection = 0) %>%
  inner_join(max) %>%
  drop_na(tstart) %>%
  bind_rows(all_events) %>%
  arrange(eid, tstart) 

z <- d %>%
  left_join(rare_var) %>%
  drop_na(rare_var_CRP) %>%
  filter(time_event > tstart)

model.1 = coxph(Surv(tstart,time_event,had_infection) ~ as.factor(rare_var_CRP) + cluster(eid),robust=TRUE, data = z)
summary(model.1)

```

Link to covars;

1.  generate cox regression df
2.  generation time to death
3.  death censor

```{r}


cox_ready <- covars %>%
  left_join(infection_events) %>%
  left_join(sepsis_events) %>%
  left_join(pneumonia_events) %>%
  replace_na(list(had_pneumonia =0, had_sepsis = 0, had_infection = 0)) %>%
   mutate(max_fu = as.numeric(difftime(ymd("2020-12-31"),study_entry,  units = "days"))) %>%
  mutate(time_to_death = as.numeric(difftime(date_of_death, study_entry, units = "days"))) %>%
  mutate(max_fu = if_else(is.na(time_to_death), max_fu, time_to_death)) %>%
    mutate(time_event_infection = if_else(is.na(time_event_infection), max_fu, time_event_infection)) %>%
  mutate(time_event_pneumonia = if_else(is.na(time_event_pneumonia), max_fu, time_event_pneumonia)) %>%
  mutate(time_event_sepsis = if_else(is.na(time_event_sepsis), max_fu, time_event_sepsis)) 
#generate death data

cox_ready <- cox_ready %>%
  mutate(pneumonia_death = case_when(
    had_pneumonia == 1 &
      time_event_pneumonia + 28 > time_to_death ~ 1,
    TRUE ~ 0
  )) %>%
   mutate(infection_death = case_when(
    had_infection == 1 &
      time_event_infection + 28 > time_to_death ~ 1,
    TRUE ~ 0
  )) %>%
   mutate(sepsis_death = case_when(
    had_sepsis == 1 &
      time_event_sepsis + 28 > time_to_death ~ 1,
    TRUE ~ 0
   )) 
  # %>%
  # mutate(time_event_infection_censored= if_else(is.na(date_of_death), time_event_infection, time_to_death)) %>%
  # mutate(time_event_sepsis_censored= if_else(is.na(date_of_death), time_event_sepsis, time_to_death)) %>%
  # mutate(time_event_pneumonia_censored= if_else(is.na(date_of_death), time_event_pneumonia, time_to_death)) 
  # 

#Generate censored death date

cox_ready <- cox_ready %>%
    mutate(time_event_death_censor = if_else(is.na(time_to_death), max_fu, time_to_death)) %>%
  filter(eid %in% prim_care_linked$eid)





```

Table 1

```{r}
cox_ready %>%
  drop_na(c_reactive_protein) %>%
  mutate(c_reactive_protein = Hmisc::cut2(c_reactive_protein, g = 4)) %>%
  select(age_at_entry, sex, history_stroke, history_MI, history_asthma, history_copd, history_dementia, history_liver, history_cancer, ever_smoked, townsend_deprivation_index_at_recruitment, body_mass_index_bmi, creatinine, c_reactive_protein) %>%
    tbl_summary(by = "c_reactive_protein") %>%
  add_p()
    

```

```{r}


p1<-cox_ready %>%
  mutate(crp = ntile(log(c_reactive_protein), 10)) %>%
  group_by(crp) %>%
     drop_na(c_reactive_protein) %>%
  group_by(crp) %>%
  summarise(median_crp = median(c_reactive_protein),
            bmean_pneumonia = mean(had_infection)) %>%
   pivot_longer(contains("mean")) %>%
   ggplot(aes(x = median_crp, y = value)) +
   geom_point() +
  geom_line() +
   theme_bw() +
   ylab("Proportion with event") +
    xlab("CRP (mg/L)") +
  xlim(0.25,9) 

p2<-cox_ready %>%
  mutate(crp = ntile(log(body_mass_index_bmi), 10)) %>%
  group_by(crp) %>%
     drop_na(c_reactive_protein) %>%
  group_by(crp) %>%
  summarise(median_crp = median(body_mass_index_bmi),
            bmean_pneumonia = mean(had_infection)) %>%
   pivot_longer(contains("mean")) %>%
   ggplot(aes(x = median_crp, y = value)) +
   geom_point() +
  geom_line() +
   theme_bw() +
   ylab("Proportion with event") +
  xlab("Body mass index (kg/m2)") 
p1/p2
```

Genetic data common

```{r}

common_final <- common_prs %>%
  left_join(final) %>%
  drop_na(crp_prs_scaled)




common_final %>%
  mutate(decile = ntile(crp_prs, 4)) %>%
  group_by(decile) %>%
  summarise(crp_m = median(c_reactive_protein, na.rm = T))


common_final %>%
  drop_na(crp_prs) %>%
  sample_n(1e5) %>%
  mutate(crp_prs = Hmisc::cut2(crp_prs, g = 4)) %>%
  select(age_at_entry, sex, history_stroke, history_MI, history_asthma, history_copd, history_dementia, history_liver, history_cancer, ever_smoked, townsend_deprivation_index_at_recruitment, body_mass_index_bmi, creatinine, c_reactive_protein) %>%
    tbl_summary(by = "crp_prs") %>%
  add_p()

    

```

Genetic models

```{r}

x <- "Surv(time_event_infection, had_infection)" 

run_table_gen <- function(x, y) {


f <- as.formula(sprintf(paste0(x, "~ crp_prs_scaled + sex + rcs(age_at_entry) + uk_biobank_assessment_centre + history_stroke + history_MI + history_copd + history_liver + history_cancer + ever_smoked + creatinine + body_mass_index_bmi ")))
m <- (coxph(formula = f, data = common_final))
m2 <- tidy(m, exponentiate = T, conf.int = T) %>% mutate(model = y) 
return(m2)
}

tab2 <- bind_rows(
run_table_gen("Surv(time_event_infection, had_infection)", "Infection"),
run_table_gen("Surv(time_event_pneumonia, had_pneumonia)", "Pneumonia"),
run_table_gen("Surv(time_event_sepsis, had_sepsis)", "Sepsis"),

run_table_gen("Surv(time_event_death_censor, infection_death)", "Infection death"),
run_table_gen("Surv(time_event_death_censor, pneumonia_death)", "Pneumonia death"),
run_table_gen("Surv(time_event_death_censor, sepsis_death)", "Sepsis death")

)



tab2 %>%
  filter(term == "crp_prs_scaled") %>%
  mutate(across(c(estimate,conf.low, conf.high, p.value),~ signif(.x, digits = 3))) %>%
  transmute(model, estimate = paste0(estimate, " (", conf.low, "-",conf.high,")"), p.value) %>%
  group_by(model) %>%
  gt::gt()
```

```{r}
rare_vs_common <- rare_var %>%
  left_join(common_final) 

cor(rare_vs_common$rare_var_CRP, rare_vs_common$crp_prs, use = "complete.obs")

m <- lm(ln_crp ~ rare_var_CRP*crp_prs, data = rare_vs_common)
tidy(m)

rare_vs_common %>%
  group_by(rare_var_CRP) %>%
  summarise(median_prs = median(crp_prs, na.rm = T))


rare_final <- rare_var %>%
  right_join(final) %>%
  drop_na(rare_var_CRP) %>%
  left_join(principle_components)


rare_final %>%
  count(rare_var_CRP) %>%
  mutate(x = n/sum(n) * 100)

rare_final %>%
  drop_na(c_reactive_protein) %>%
  group_by(rare_var_CRP) %>%
  summarise(median = median(c_reactive_protein, na.rm = T),
            upper = quantile(c_reactive_protein, 0.75),
            lower = quantile(c_reactive_protein, 0.25), n = n())

rare_final %>%
  mutate(rs77832441_A = if_else(rare_var_CRP == 1, "Carrier", "Non Carrier")) %>%
  select(eid, rs77832441_A, c_reactive_protein) %>%
  ggplot(aes(x = c_reactive_protein)) +
  geom_density(aes(group = rs77832441_A, color = rs77832441_A)) +
  scale_x_log10() +
  xlab("CRP (mg/L)") +
  theme_bw()
```

```{r}

rare_final %>%
   count(rare_var_CRP)
# 
  # rare_final <- rare_final %>%
  #  filter(rare_var_CRP != 2 )

x <- "Surv(time_event_infection, had_infection)" 

run_table_gen <- function(x, y) {


f <- as.formula(sprintf(paste0(x, "~ rare_var_CRP + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10")))
m <- (coxph(formula = f, data = rare_final))
m2 <- tidy(m, exponentiate = T, conf.int = T) %>% mutate(model = y) 
return(m2)
}

tab2 <- bind_rows(
run_table_gen("Surv(time_event_infection, had_infection)", "Infection"),
run_table_gen("Surv(time_event_pneumonia, had_pneumonia)", "Pneumonia"),
run_table_gen("Surv(time_event_sepsis, had_sepsis)", "Sepsis"),

run_table_gen("Surv(time_event_death_censor, infection_death)", "Infection death"),
run_table_gen("Surv(time_event_death_censor, pneumonia_death)", "Pneumonia death"),
run_table_gen("Surv(time_event_death_censor, sepsis_death)", "Sepsis death")

)



tab2 %>%
  filter(term == "rare_var_CRP") %>%
  mutate(across(c(estimate,conf.low, conf.high, p.value),~ signif(.x, digits = 3))) %>%
  transmute(model, estimate = paste0(estimate, " (", conf.low, "-",conf.high,")"), p.value) %>%
  group_by(model) %>%
  gt::gt()


x <- "Surv(time_event_pneumonia, had_pneumonia)"
x <- "Surv(time_event_infection, had_infection)"
f <- as.formula(sprintf(paste0(x, "~ rare_var_CRP")))
m <- (coxph(formula = f, data = rare_final))
p <- survfit(Surv(time_event_infection, had_infection)~ rare_var_CRP, data = rare_final)
ggsurvplot(p, data = rare_final, pval = T, xlim = c(0,5000))
p
rare_final %>%
  count(rare_var_CRP, had_)


rare_final <- rare_final %>%
  mutate(rs77832441 =if_else(rare_var_CRP ==1, "AG", "GG"))

rare_final <- rare_final %>%
  mutate(rs77832441 =rare_var_CRP)

fit <- survfit(Surv(time_event_infection, had_infection) ~ rs77832441, data = rare_final)
fit
ggsurvplot(fit, data = rare_final, conf.int = T, ylim = c(0.6,1), pval = T)

indiv <- rare_final %>%
  filter(had_sepsis == 1) %>%
  mutate(time = if_else(
    time_event_death_censor - time_event_sepsis > 28, 28,time_event_death_censor - time_event_sepsis 
  ))

indiv %>%
  filter(rare_var_CRP == 1) %>%
  filter(time <28)

fit <- survfit(Surv(time, sepsis_death) ~ rare_var_CRP, data = indiv)
ggsurvplot(fit, data = indiv, conf.int = T)

d <- rare_final %>%
  mutate(dead = if_else(time_to_death < max_fu,1,0)) 


```

# MR

```{r}


x <- vroom("https://github.com/gushamilton/il6-sepsis/raw/main/data/final_mr_results.tsv")
x<- vroom("https://raw.githubusercontent.com/gushamilton/il6-sepsis/main/data/harmonised_data_final.tsv")%>%
  mutate(beta.exposure = if_else(SNP == "rs3093077", -beta.exposure, beta.exposure)) %>%
  mr(method = "mr_ivw")
conversions <- read_csv("~/R/il6-toci/conversions.csv")

ivw <- function(df) {
  TE = df$b
  SE = df$se
  m1 <- metagen(TE, SE) 
  tibble(estimate = m1$TE.fixed,
         std.error= m1$seTE.fixed,
         het = m1$pval.Q,
         pval = m1$pval.fixed,
         lower = m1$lower.fixed,
         upper = m1$upper.fixed)
  
}


p1 <- x %>%
  filter(exposure == "cisCRP") %>%
  filter(method == "Inverse variance weighted") %>%
  left_join(conversions) %>%
  group_by(name) %>%
  nest() %>%
  mutate(meta_an = map(data, ivw)) %>%
  unnest(meta_an) %>%
  drop_na(het) %>%
  ungroup() %>%
  select(name, b = estimate, se = std.error, pval)  %>%
  mutate(b = -b) %>%
  arrange(-b) %>%
  ggforestplot::forestplot(name = name, estimate = b, se = se, logodds = T) +
  theme_bw() +
  xlab("Odds ratio (95% CI)")

p2 <- x %>%
  filter(exposure == "cisCRP") %>%
  filter(method == "Inverse variance weighted") %>%
  left_join(conversions) %>%
  filter(group == "gains") %>%
 filter(!str_detect(name, "CAP")) %>%
  mutate(b = -b) %>%
  arrange(-b) %>%
  ggforestplot::forestplot(name = name, estimate = b, se = se, logodds = T) +
  theme_bw() +
  xlab("Odds ratio (95% CI)")


p3 <- x %>%
  filter(exposure == "cisCRP") %>%
  filter(method == "Inverse variance weighted") %>%
  left_join(conversions) %>%
  filter(group == "crit") %>%
  filter(!str_detect(name, "CAP")) %>%
  mutate(b = -b) %>%
  arrange(-b) %>%
  ggforestplot::forestplot(name = name, estimate = b, se = se, logodds = T) +
  theme_bw() +
  xlab("Odds ratio (95% CI)")

((p1 | p3) + plot_annotation(tag_levels = "A"))


x %>%
  filter(exposure == "cisCRP") %>%
  filter(method == "Inverse variance weighted") %>%
  left_join(conversions) %>%
    mutate(b = -b) %>%
  group_by(name) %>%
  nest() %>%
  mutate(meta_an = map(data, ivw)) %>%
  unnest(meta_an) %>%
  drop_na(het) %>%

  ungroup() %>%
  mutate(across(c(estimate,lower, upper), exp)) %>%
  view()
 
 x %>%
  filter(exposure == "cisCRP") %>%
  filter(method == "Inverse variance weighted") %>%
  left_join(conversions) %>%
  filter(group == "crit") %>%
  filter(!str_detect(name, "CAP")) %>%
  mutate(b = -b) %>%
  mutate(upper = exp(b-1.96*se), lower = exp(b+1.96*se), b = exp(b)) %>% view()

```

##PHEWAS

```{r}
phewas <- vroom("~/data/CRP/results-combined.txt.gz")


phewas <- phewas %>%
  mutate(n_case = sub(".*/", "", n)) %>%
  mutate(n_case = as.numeric(sub("\\(.*", "", n_case)), n_control = as.numeric(sub("\\/.*", "", n))) %>%
   filter(n_case >100) %>%
   filter(Cat1_Title == "Biological samples" |
          Cat3_Title == "First occurrences" |
          Cat2_Title == "Physical measures")


phewas <- phewas %>%
  mutate(pvalue = if_else(pvalue == 0, 1e-320, pvalue))
qqman::qq(phewas$pvalue)

phewas$resType
final <-phewas %>%
  filter(abs(beta) < 900) %>%
  select(resType,description, beta, lower, upper, pvalue, n_case, n_control) %>%
    mutate(se = (upper - beta) / 1.96) %>%
  mutate(pval_FDR = p.adjust(pvalue, method = "fdr")) %>%
    mutate(pval_bonferonni = p.adjust(pvalue, method = "bonferroni")) %>%
  mutate(description = str_remove_all(description, "Date |first reported|\\(|\\)")) %>%
  mutate(across(c(beta,lower,upper), 
                ~if_else(str_detect(resType, "LOGIS"), exp(.x), .x)
                
                )) %>%
  mutate(across(c(beta,lower,upper, pval_bonferonni, pval_FDR, pvalue, se), signif, digits = 3)) 


x <- c("I21", "stroke", "HDL", "Trigl", "Body", "Syst")

final %>%
  arrange(resType) %>%
  filter(str_detect(description, paste(x, collapse = "|"))) %>%
  mutate(beta = paste0(beta, " (",lower,",", upper, ")")) %>%
  select(-c(upper,lower, se, pval_FDR, pval_bonferonni, resType)) %>%
  gt::gt()


```
